src/xml_parser_main_metadata_together.c

```diff
@@ -334,136 +328,227 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
                                         void *pkgcb_data,
                                         cr_XmlParserWarningCb warningcb,
                                         void *warningcb_data,
-                                        gboolean allow_out_of_order,
                                         GError **err)
 {
-    int ret = CRE_OK;
-    CR_FILE *primary_f = NULL;
-    CR_FILE *filelists_f = NULL;
-    CR_FILE *other_f = NULL;
-    GError *tmp_err = NULL;
+    assert(pkgcb || newpkgcb);
+    cr_PkgIterator* pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, newpkgcb, newpkgcb_data, warningcb, warningcb_data, err
+    );
 
-    cr_CbData cbdata;
-    cbdata.in_progress_pkgs_list = NULL;
-    cbdata.in_progress_pkgs_hash = NULL;
-    cbdata.newpkgcb = newpkgcb;
-    cbdata.newpkgcb_data = newpkgcb_data;
-    cbdata.pkgcb = pkgcb;
-    cbdata.pkgcb_data = pkgcb_data;
+    assert(pkg_iterator);
+    if (*err) {
+        return (*err)->code;
+    }
 
-    if (allow_out_of_order) {
-        cbdata.in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    cr_Package* package = NULL;
+    GError* tmp_err = NULL;
+
+    while (package = cr_PkgIterator_parse_next(pkg_iterator, err)) {
+        if (pkgcb) {
+            // call user package callback
+            // pkgcb() destroys the package!!
+            if (pkgcb(package, pkgcb_data, &tmp_err)) {
+                // Error condition
+                if (tmp_err) {
+                    g_propagate_prefixed_error(err, tmp_err, "Parsing interrupted: ");
+                } else {
+                    g_set_error(err, ERR_DOMAIN, CRE_CBINTERRUPTED, "Parsing interrupted");
+                }
+                cr_PkgIterator_free(pkg_iterator, err);
+                return CRE_CBINTERRUPTED;
+            } else {
+                // If callback return CRE_OK but it simultaneously set
+                // the tmp_err then it's a programming error.
+                assert(tmp_err == NULL);
+            }
+        } else {
+            // Free the package if there is no external callback to do so.
+            cr_package_free(package);
+        }
     }
 
+    assert(cr_PkgIterator_is_finished(pkg_iterator));
```

Comment:
I'm just rebuilding via the Python package, and I believe that always (or at least by default) builds w/ release configuration, so that's probably true.